packages/client/src/rpc/modules/debug.ts

```diff
@@ -170,4 +184,98 @@ export class Debug {
       }
     }
   }
+
+  /**
+   * Returns a trace of an eth_call within the context of the given block execution using the final state of the parent block
+   * @param params an array of 3 parameters:
+   *    1. an {@link RpcTx} object that mirrors the eth_call parameters object
+   *    2. A block hash or number formatted as a hex prefixed string
+   *    3. An optional tracer options object
+   * @returns an execution trace of an {@link eth_call} in the context of a given block execution
+   * mirroring the output from {@link traceTransaction}
+   */
+  async traceCall(params: [RpcTx, string, tracerOpts]) {
+    const [callArgs, blockOpt, tracerOpts] = params
+    // Validate configuration and parameters
+    if (!this.service.execution.receiptsManager) {
+      throw {
+        message: 'missing receiptsManager',
+        code: INTERNAL_ERROR,
+      }
+    }
+
+    if (this.vm === undefined) {
+      throw new Error('missing vm')
+    }
+
+    const opts = validateTracerConfig(tracerOpts)
+
+    const block = await getBlockByOption(blockOpt, this.chain)
+    const parentBlock = await this.service.chain.getBlock(block.header.parentHash)
+
+    const vm = await this.vm.shallowCopy()
```

Comment:
This has an underlying Trie which writes to disk - right? I don't think we want that, we want that the Trie is in memory for these calls, otherwise we clog the disk with calls which are possibly never made in the chain. (Also a DoS vector since you can spam the RPC with random txs and thus hog disk space)

Reply:
We need to operate on the same datastructure (and actually: state) and it is not possible/practical to copy over the state to memory on these kind of instantiations.

So I would say on first round memory read is not possible. This *does* use the statemanager + all initialized caches though, so depending on the request this might not need direct trie/state reads.

That said: If I am not mistaken my understanding of RPC (as it is designed for Ethereum) is that it is *always* a potential DoS surface you cannot hide from, and therefore it is just not recommended to completely expose an RPC interface in non-testing/devnet situations.

Calling RPC methods just *does* use additional ressources, and calling them often uses even more. 😋

There are likely more general solutions and mitigations - e.g. block certain request IPs after x calls or after ressouce-intensive calls. Not sure if this is so useful for us atm with no real production users.

Reply:
This is why we need to implement dynamic snapshots! 😀